### PR TITLE
Update FDA plotting and fitting

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
**Description of work.**

The frequency domain analysis GUI has numerous usability problems. This PR fixes enough of the issues that it is possible to perform transforms on time domain data and then perform fits on the transformed values but there is more work still to be done.

**To test:**
 * Test that you can perform frequency domain transforms and perform fits on single data workspaces. Check that the plots update in the expected manner when you do so. Get in touch with me to explain the workflows in this GUI is you are not already familiar with it.

Fixes 28742 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
This is a nice set of changes for both GUIs, it was a good idea to introduce PlotMode. Just a few comments and one issue:

For me the FDA fitting isn't working, it complains about there not being a fit function defined. I think this is partly due to me. To support two tabs accessing the same model and allow any row of the sequential table to be fitting individually, I created a workspace fit function [map](https://github.com/mantidproject/mantid/blob/359994311ec8c708a0c2406b70de9d8d243e5706/scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py#L444). I think its not creating this workspace map correctly for FDA, as the keys of the map are:

`[frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'}), frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'}), frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'})]`

Whereas the input workspace to the fit is 'FFT; Re MUSR62260; Group; fwd; Asymmetry; FD; Im MUSR62260; Group; fwd; Asymmetry; FD_mod', which means it cannot find the correct function (as it usees this workspace name to grab the correct fit function).

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_view.py [hunk 1]
```diff
@@ -120,3 +138,41 @@ def set_plot_type(self, plot_type: str):
...
+        """
+        self.plot_type_combo.setEnabled(False)
+
+    def enable_tile_plotting_options(self):
+        """
+        Enable tile plotting
+        """
+        self.tiled_plot_checkbox.setEnabled(True)
+        self.tiled_by_combo.setEnabled(True)
+        self.plot_raw_checkbox.setEnabled(True)
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/plot_widget/plot_widget_view.py:170]: Good point

**Comment by Mantid-Matthew** [comment]
> 
> 
> This is a nice set of changes for both GUIs, it was a good idea to introduce PlotMode. Just a few comments and one issue:
> 
> For me the FDA fitting isn't working, it complains about there not being a fit function defined. I think this is partly due to me. To support two tabs accessing the same model and allow any row of the sequential table to be fitting individually, I created a workspace fit function [map](https://github.com/mantidproject/mantid/blob/359994311ec8c708a0c2406b70de9d8d243e5706/scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_model.py#L444). I think its not creating this workspace map correctly for FDA, as the keys of the map are:
> 
> `[frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'}), frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'}), frozenset({'MUSR62260; Group; fwd; Asymmetry; FD'})]`
> 
> Whereas the input workspace to the fit is 'FFT; Re MUSR62260; Group; fwd; Asymmetry; FD; Im MUSR62260; Group; fwd; Asymmetry; FD_mod', which means it cannot find the correct function (as it usees this workspace name to grab the correct fit function).

Yes you are entirely right here. I've now added a fix which I think will get the fitting to work for the freuency domain analysis. I suspect this will need to be further modified if batch fitting is ever added to the FDA GUI. But as that may not ever happen and I am not quite sure of what form it will take decided not to complicate things further for now.

**Comment by Mantid-Matthew** [comment]
@StephenSmith25 Just remembered that I was also planning to disable the Ts Asymmetry mode option in the FDA GUI so hold off on looking at this again until I have done that.

**Comment by StephenSmith25** [comment]
This works well now, just one thing which you may want to change. The `Time Start` and `Time end` boxes in the fitting tab should probably be renamed to something more appropriate for FDA.

**Comment by Mantid-Matthew** [comment]
> This works well now, just one thing which you may want to change. The `Time Start` and `Time end` boxes in the fitting tab should probably be renamed to something more appropriate for FDA.

Yes good spot and given some more appropriate initial values as well.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_view.py [hunk 2]
```diff
@@ -35,6 +35,11 @@ def __init__(self, parent=None):
 
         self.disable_simul_fit_options()
 
+        if is_frequency_domain:
+            self.hide_simultaneous_fit_options()
+            self.fit_options_table.hideRow(3)
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_view.py:40]: what are these magic numbers?

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py [hunk 3]
```diff
@@ -11,16 +11,21 @@
...
     PlottingCanvasPresenterInterface
+from Muon.GUI.Common.contexts.frequency_domain_analysis_context import FrequencyDomainAnalysisContext
 from Muon.GUI.Common.plot_widget.plot_widget_model import PlotWidgetModel
 from Muon.GUI.Common.plot_widget.plot_widget_view_interface import PlotWidgetViewInterface
+from Muon.GUI.Common.contexts.muon_gui_context import PlotMode
 from mantidqt.utils.observer_pattern import GenericObserver, GenericObserverWithArgPassing, GenericObservable
 from mantid.dataobjects import Workspace2D
 
+MUON_ANALYSIS_DEFAULT_X_RANGE = [0.0, 15.0]
+FREQUENCY_DOMAIN_ANALYSIS_DEFAULT_X_RANGE = [0.0, 1000.0]
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:22]: should we get these from the context?

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py [hunk 4]
```diff
@@ -83,7 +94,73 @@ def handle_data_updated(self):
...
+    def update_plot(self, autoscale=False):
+        if self.context.gui_context['PlotMode'] == PlotMode.Data:
+            self.handle_data_updated(autoscale=autoscale)
+        elif self.context.gui_context['PlotMode'] == PlotMode.Fitting:  # Plot the displayed workspace
+            self.handle_plot_selected_fits(
+                self._get_selected_fit_workspaces(), autoscale
+            )
+
+    def handle_plot_mode_changed(self, plot_mode : PlotMode):
+        if isinstance(self.context, FrequencyDomainAnalysisContext):
```
Comment (AnthonyLim23) [inline scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:108]: I have seen this a few times. Would it be worth making it a method of the context?

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py (hunk 3 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:22]: Yes good idea that removes the need for one of the if statements.

File: scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_view.py (hunk 2 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/fitting_tab_widget/fitting_tab_view.py:40]: These are refencing the names of the table displaying information. I've changed them to named constants to make this clearer.

File: scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py (hunk 4 above)
Comment (Mantid-Matthew) [reply scripts/Muon/GUI/Common/plot_widget/plot_widget_presenter.py:108]: I don't think you gain much by adding a method to the context identifying what type of context it is when python as a language allows you to easily check the type of objects. The ideally we should be redeucing the number of these checks going forward as well as we differentiate between the different GUI's by more clearly seperating there code into different classes.

**Comment by AnthonyLim23** [comment]
clicking `calculate FFT` causes an error `QComboBox object has no attribute setCurrentText`

**Comment by AnthonyLim23** [comment]
the xrange doesnt seem to change to frequency when plotting it

**Comment by AnthonyLim23** [comment]
the plots shown seem to be somewhat random. If you do a default FFT on MUSR62260, then change the Im group to something else. The new FFT will be plotted on only some plots.

**Comment by AnthonyLim23** [comment]
doing a fit in the time domain GUI throws a long stack trace. Seems to be from the ADS observer and might be because I have loaded the same data into FDA first.

**Comment by Mantid-Matthew** [comment]
> 
> 
> the xrange doesnt seem to change to frequency when plotting it outside of fit

Good spot, I was planning to remove the ability to plot frequency outside of fits but forgot

**Comment by Mantid-Matthew** [comment]
> 
> 
> clicking `calculate FFT` causes an error `QComboBox object has no attribute setCurrentText`

Looks like setCurrentText became deprecated at some point and has presumably dissapeared on the version of Qt that you are using. I'll use something else. Out of interest which system/qt version are you testing on?

**Comment by Mantid-Matthew** [comment]
> 
> 
> the plots shown seem to be somewhat random. If you do a default FFT on MUSR62260, then change the Im group to something else. The new FFT will be plotted on only some plots.

Could you elaborate on how to do this? For me it always plots the mod of the latest FFT done.

**Comment by StephenSmith25** [comment]
> doing a fit in the time domain GUI throws a long stack trace. Seems to be from the ADS observer and might be because I have loaded the same data into FDA first.

I've noticed this as well. If you've had both GUIS open the ADS observers seem to conflict. Think this needs to be looked into.

**Comment by Mantid-Matthew** [comment]
> 
> 
> doing a fit in the time domain GUI throws a long stack trace. Seems to be from the ADS observer and might be because I have loaded the same data into FDA first.

I am struggling to reproduce this. Can you remember the precise order of steps which produced the stacktrace? Also could you send me the stacktrace.

**Comment by AnthonyLim23** [comment]
> > clicking `calculate FFT` causes an error `QComboBox object has no attribute setCurrentText`
> 
> Looks like setCurrentText became deprecated at some point and has presumably dissapeared on the version of Qt that you are using. I'll use something else. Out of interest which system/qt version are you testing on?

I think its Qt4, Im using MantidPlot on Windows (get an error trying to run workbench thats unrelated)

**Comment by AnthonyLim23** [comment]
> > the xrange doesnt seem to change to frequency when plotting it outside of fit
> 
> Good spot, I was planning to remove the ability to plot frequency outside of fits but forgot. Do we want to disallow looking at frequency in data mode or add in the scale chaneing?

we want transform tab to show freq data. But can do that by switching to data

**Comment by Mantid-Matthew** [comment]
In summary I have now:
  * Corrected the issue with it not working in qt4
  * Removed the ability to perform frequency domain plots from the data plots for now
  * The issue with plots not updating when performing new transformed is solved by point 1 here.
  * We've agreed to leave the ADS observer issue for another PR

So should be good for a re-look

**Comment by AnthonyLim23** [comment]
flake8 errors

**Comment by AnthonyLim23** [comment]
Do an FFT and the fit to the data.
Go to the results tab and it has been added
Go back to fit tab
Fit some different data 
Go to the results tab, the data has not been added

**Comment by Mantid-Matthew** [comment]
> Do an FFT and the fit to the data.
> Go to the results tab and it has been added
> Go back to fit tab
> Fit some different data
> Go to the results tab, the data has not been added

This issues predates this PR so as this has gotten quite large I have added a new issue for this #29015

**Review by AnthonyLim23 — APPROVED** [review]

**Review by StephenSmith25 — APPROVED** [review]

**Comment by AnthonyLim23** [comment]
@mantidproject/gatekeepers could we please merge this?

**Pull request merged**